File: CSharp/SpaceVIL/SpaceVIL/Items/ItemText.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace SpaceVIL
{
    abstract class ItemText : Primitive
    {
        private List<float> _alphas;
        private float[] _coordinates;
        private float[] _colors;
        private String _itemText = "";

        private Font _font = new Font(new FontFamily("Courier New"), 22, FontStyle.Regular);

        private bool _criticalFlag = true;
        private bool _coordsFlag = true;
        private bool _colorFlag = true;

        public ItemText()
        {
            SetBackground(Color.Black);
            SetWidthPolicy(SizePolicy.Expand);
            SetHeightPolicy(SizePolicy.Expand);
        }

        public ItemText(String text, Font font, String name) : this()
        {
            _itemText = text;
            this._font = font;
        }

        protected void SetRealCoords(List<float> realCoords)
        {
            _coordinates = ToGL(realCoords);
        }

        protected void SetAlphas(List<float> alphas)
        {
            this._alphas = alphas;
            SetColor(alphas);
        }

        internal String GetItemText()
        {
            return _itemText;
        }
        internal void SetItemText(String itemText)
        {
            if (!this._itemText.Equals(itemText))
            {
                this._itemText = itemText;
                _criticalFlag = true;
            }
        }

        internal Font GetFont()
        {
            return _font;
        }
        internal void SetFont(Font font)
        {
            if (!this._font.Equals(font))
            {
                this._font = font;
                _criticalFlag = true;
            }
        }

        public abstract void UpdateData(UpdateType updateType);

        internal float[] GetCoordinates()
        {
            if (_criticalFlag)
            {
                UpdateData(UpdateType.Critical);
                _criticalFlag = false;
                _coordsFlag = false;
            }
            else if (_coordsFlag)
            {
                UpdateData(UpdateType.CoordsOnly);
                _coordsFlag = false;
            }
            return _coordinates;
        }

        internal float[] GetColors()
        {
            if (_colorFlag)
            {
                SetAlphas(_alphas);
                _colorFlag = false;
            }
            return _colors;
        }

        private float[] ToGL(List<float> coord)
        {
            float[] outCoord = new float[coord.Count];
            float f;
            float x0 = GetX();
            float y0 = GetY();
            float windowH = GetHandler().GetHeight() / 2f;
            float windowW = GetHandler().GetWidth() / 2f;

            float xmin = Int32.MaxValue;
            float xmax = Int32.MinValue;
            float ymin = Int32.MaxValue;
            float ymax = Int32.MinValue;

            for (int i = 0; i < coord.Count; i += 3)
            {
                f = coord[i];
                xmin = (xmin > f) ? f : xmin;
                xmax = (xmax < f) ? f : xmax;
                f += x0;
                f = f / windowW - 1.0f;
                outCoord[i] = f;

                f = coord[i + 1];
                ymin = (ymin > f) ? f : ymin;
                ymax = (ymax < f) ? f : ymax;
                f += y0;
                f = -(f / windowH - 1.0f);
                outCoord[i + 1] = f;

                f = coord[i + 2];
                outCoord[i + 2] = f;
            }

            SetWidth((int)Math.Abs(xmax - xmin));
            SetHeight((int)Math.Abs(ymax - ymin));

            return outCoord;
        }

        private Color _foreground = Color.White; //default
        public Color GetForeground()
        {
            return _foreground;
        }
        public void SetForeground(Color foreground)
        {
            if (!this._foreground.Equals(foreground))
            {
                this._foreground = foreground;
                _colorFlag = true;
            }
        }

        private void SetColor(List<float> alphas)
        {
            _colors = new float[alphas.Count * 4];
            
            Color col = GetForeground();
            float r = col.R * 1f / 255f;
            float g = col.G * 1f / 255f;
            float b = col.B * 1f / 255f;

            int inc = 0;
            foreach (float f in alphas)
            {
                _colors[inc] = r; inc++;
                _colors[inc] = g; inc++;
                _colors[inc] = b; inc++;
                _colors[inc] = f; inc++;
            }
        }

        private ItemAlignment _textAlignment;
        public ItemAlignment GetTextAlignment()
        {
            return _textAlignment;
        }
        public void SetTextAlignment(ItemAlignment value)
        {
            //Bottomif (!_textAlignment.Equals(value))
            {
                _textAlignment = value;
                _coordsFlag = true;
            }
        }

        protected void SetCoordsFlag(bool flag)
        {
            _coordsFlag = flag;
        }

        public virtual float[] Shape()
        {
            return GetCoordinates();
        }
    }
}

File: Java/Fonts/ItemText.java
package CommonVesselGUI;

import java.awt.*;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public abstract class ItemText extends Primitive {
    //private static int count = 0;
    //private List<Float> realCoords;
    private List<Float> alphas;
    private float[] coordinates;
    private float[] colors;
    private FontEngine fontEngine;
    private String itemText;

    private Font font;

    private boolean criticalFlag = true;
    private boolean coordsFlag = true;
    private boolean colorFlag = true;

    private double alignShiftX = 0;
    private double alignShiftY = 0;

    private float textWidth;
    private float textMinY;
    private float textMaxY;

    ItemText(WindowLayout wnd_handler, String text, Font font, String name) {
        super(wnd_handler, name); //"Text" + count);
        //count++;
        //default
        itemText = text;
        this.font = font;

        //super.setXPosition(100);
        //super.setYPosition(300); //TODO это нижняя левая точка, но не совсем. См. l0.0188.0.

        //setBackground(new Color(255, 255, 255, 255));

        setWidthPolicy(SizePolicy.Fixed);
        setHeightPolicy(SizePolicy.Fixed);
        //----------------------------------------------------------------------------------------

        //updateMainArrays();
    }

    protected void setRealCoords(List<Float> realCoords) {
        coordinates = toGL(realCoords);
    }

    protected void setAlphas(List<Float> alphas) {
        this.alphas = alphas;
        colors = setColor(alphas);
    }

    String getItemText() { return itemText; }
    void setItemText(String itemText) {   //TODO update
        if (!this.itemText.equals(itemText)) {
            this.itemText = itemText;
            criticalFlag = true;
        }
    }

    Font getFont() { return font; }
    void setFont(Font font) {   //TODO update
        if (!this.font.equals(font)) {
            this.font = font;
            criticalFlag = true;
        }
    }

    @Override
    protected void invokePoolEvents() {

    }

    abstract void updateData(UpdateType updateType);

    float[] getCoordinates() {
        if (criticalFlag) {
            updateData(UpdateType.Critical);
            criticalFlag = false;
            coordsFlag = false;
            colorFlag = false;
        }
        else if (coordsFlag) {
            updateData(UpdateType.CoordsOnly);
            coordsFlag = false;
            colorFlag = false;
        }
        else if (colorFlag) {
            setColor(alphas);
            colorFlag = false;
        }
        return coordinates;
    }

    float[] getColors() {
        if (criticalFlag) {
            updateData(UpdateType.Critical);
            criticalFlag = false;
            coordsFlag = false;
            colorFlag = false;
        }
        else if (coordsFlag) {
            updateData(UpdateType.CoordsOnly);
            coordsFlag = false;
            colorFlag = false;
        }
        else if (colorFlag) {
            setColor(alphas);
            colorFlag = false;
        }
        return colors;
    }

    private float[] toGL (List<Float> coord) {
        float[] outCoord = new float[coord.size()];
        float f;
        float x0 = getXPosition();
        float y0 = getYPosition();
        float windowH = getParent().getHeight() / 2f; //wnd.getHeight() / 2;
        float windowW = getParent().getWidth() / 2f; //wnd.getWidth() / 2;
        float xmin = coord.get(0);
        float xmax = coord.get(0);
        float ymin = coord.get(1);
        float ymax = coord.get(1);

        for (int i = 0; i < coord.size(); i += 3) {
            f = coord.get(i) ;
            xmin = (xmin > f) ? f : xmin;
            xmax = (xmax < f) ? f : xmax;
            f += x0;
            f = f / windowW - 1.0f;
            outCoord[i] = f;

            f = coord.get(i + 1);
            ymin = (ymin > f) ? f : ymin;
            ymax = (ymax < f) ? f : ymax;
            f += y0;
            f = -( f / windowH - 1.0f);
            outCoord[i + 1] = f;

            f = coord.get(i + 2);
            outCoord[i + 2] = f;
        }

        setWidth((int)Math.abs(xmax - xmin)); //???
        setHeight((int)Math.abs(ymax - ymin));

        return outCoord;
    }

    private Color foreground = new Color(255, 255, 255, 255); //default
    public Color getForeground() { return foreground; }
    public void setForeground(Color foreground) {
        if (!this.foreground.equals(foreground)) {
            this.foreground = foreground;
            colorFlag = true;
        }
    }

    private float[] setColor(List<Float> alphas) {
        float[] out = new float[alphas.size() * 4];

        Color col = getForeground();
        float r = col.getRed() * 1f / 255f;
        float g = col.getGreen() * 1f / 255f;
        float b = col.getBlue() * 1f / 255f;

        int inc = 0;
        for (float f : alphas) {
            out[inc] = r; inc++;
            out[inc] = g; inc++;
            out[inc] = b; inc++;
            out[inc] = f; inc++;
        }

        return out;
    }

    private List<ItemAlignment> textAlignment = new LinkedList<>();
    public List<ItemAlignment> getTextAlignment() { return textAlignment; }
    public void setTextAlignment(ItemAlignment... value) {
        List<ItemAlignment> list = Arrays.stream(value).collect(Collectors.toList());
        if (!textAlignment.equals(list)) {
            textAlignment = list;
            coordsFlag = true;
        }
    }

    protected void setCoordsFlag(boolean flag) { coordsFlag = flag; }

    enum UpdateType {
        Critical,
        CoordsOnly
    }
}
